Happychat: Add notification sound for new messages

jordwest opened the pull request:
Plays a sound each time a message is received from the operator.

Tested on:
 - [x] Mac/Chrome 55
 - [x] Mac/Safari 10
 - [x] Mac/Firefox
 - [x] Win/Chrome
 - [x] Win/Firefox

### How to test

1. Check out the `add/happychat/notification-sound` branch
1. Visit http://calypso.localhost:3000/help/contact, logged in as a user with Live Chat enabled
1. Start a chat (You may need to log in to the staging happychat HUD and make yourself available as an operator)
1. Make sure your sound is on :)
1. When the operator sends a message, you should hear a sound

jordwest commented:
Hmm, just noticed this fails to play the sound when the chat is minimized

jordwest commented:
Fixed, sounds now play whether minimized or not

beaucollins commented:
You can also test this via the calyplo.live using https://happychat-io-staging.go-vip.co as an operator.

beaucollins commented on client/state/happychat/selectors.js:
```diff
@@ -1,7 +1,7 @@
  * External dependencies
  */
-import map from 'lodash/map';
+import { map, findLast } from 'lodash';
```
I think we prefer the import per function for lodash in this project.

dmsnell replied:
not in Calypso. we have a Babel transform which separates out lodash so we actually prefer using this named import syntax with multiple functions in one line

beaucollins replied:
Oh good. I was tired of single imports. Thanks for clarifying.

beaucollins commented:
Tested
- [x] Windows 10 with Edge - Works
- [ ] Windows 10 with IE - Tested but no sound plays

omarjackman commented on client/components/sound/index.jsx:
```diff
@@ -0,0 +1,44 @@
+ */
+import { Component, PropTypes } from 'react';
+
+const hasBrowserSupport = () => typeof Audio === 'function';
```
Does this need to be a function? Do we expect a late polyfill for `Audio`?

omarjackman commented:
I wonder if playing the sound via some middleware would work out better than a component. @dmsnell I know you did recent work with middleware and I'm wondering if you have any thoughts on this.

dmsnell commented:
thanks for the ping @omarjackman - this is exactly the kind of thing which should be happening in middleware and with that design all a component need do is send an action. note that passing a simple data object as a Redux action lets us choose to interpret that action however we want in the middleware stage, makes for easy testing

```js
{
	type: AUDIBLE_NOTIFICATION,
	sound: 'chat_pling',
}
```

if there is anything I can to do help with this or discourage us from merging a bell as a component please let me know; I'd be happy to help!

jordwest commented:
Thanks for checking in @omarjackman and @dmsnell. There are a couple of reasons I chose to use a component, and if middleware can solve these more nicely I'm happy to change it:

 - The component manages preloading of the sound file. Because the component is attached to a happychat component, the sound isn't loaded until the user starts a chat, avoiding every user loading the sound unnecessarily (although olark currently does). Additionally, the sound gets garbage collected once the component is unmounted.

 - No need for additional actions. To trigger an (eg `AUDIBLE_NOTIFICATION`) action, we'd need to observe the events coming in from the server to then fire off a new action. Alternatively, the middleware could just listen in to the `HAPPYCHAT_RECEIVE_EVENT` action and play the sound then. Using a component means we can just observe changes to the `timeline` property on the state tree.

dmsnell commented:
> The component manages preloading of the sound file. Because the component is attached to a happychat component, the sound isn't loaded until the user starts a chat, avoiding every user loading the sound unnecessarily (although olark currently does). Additionally, the sound gets garbage collected once the component is unmounted. We could add logic for that in Redux, but the React mounting/unmounting takes care of it nicely.

this is not a problem. this is also what the query components are there for.

```js
<QueryAudioFile src="…" />
```

then…

```js
export class QueryAudioFile extends Component {
	componentDidMount() {
		this.props.request( this.props.src )
	}

	componentWillUpdate( nextProps ) {
		this.props.release( this.props.src )
		this.props.request( nextProps.src )
	}

	componentWillUnmount() {
		this.props.release( this.props.src )
	}

	render() {
		return null;
	}
}

const mapDispatchToProps = dispatch => ( {
	release: src => dispatch( { type: AUDIO_FILE_RELEASE, src } ),
	request: src => dispatch( { type: AUDIO_FILE_REQUEST, src } ),
} )

export default connect( null, mapDispatchToProps )( QueryAudioFile )
```

at this point you can probably see how the middleware can keep its own store of which files are needed, which files are already loaded, and which files it can delete. testing is trivial too.

you bring up a good point (and this is exactly how the API middleware should function) that the middleware _knows_ that _if a message comes in for happy chat_ then _we should sound the bell_ and that behavior doesn't need to be encoded into the components. I don't this separation is absolutely necessary though because I think it's possible that in cases the behavior may be more tied to the component than the system. doing it in middleware however means that we can ring the bell no matter if the component is mounted or not.

the code here is for conceptual purposes only and should not be viewed as HOW IT NEEDS TO BE DONE :wink:

dmsnell commented:
> Using a component means we can just observe changes to the happychat.timeline property on the state tree.

not sure what you mean here. the code in this PR appears to rely on changing a property to fire off a bell, but if I understand properly, that is very little different than watching changing props to fire off some action

jordwest commented:
Hmm, wouldn't this be adding state to middleware - in the sense that its storing the audio objects? Since the underlying audio is just loaded via the HTML `<audio>` element, it seems like something well suited to a React component, no?

Another option is that we could store it in a separate singleton instance and just load the sounds as they're needed (forgetting about releasing). This package uses a similar pattern but loads on startup: https://github.com/joshwcomeau/redux-sounds

dmsnell commented:
> Hmm, wouldn't this be adding state to middleware

Yes. But it's not state that the app relies on. It doesn't have an impact on the visual look of the app and if it were purged we should be able to automatically rebuild it. This is the cornered off location _for_ stateful operations (such as "only make this API request if we haven't made the same one within the past minute")

> Since the underlying audio is just loaded via the HTML <audio> element

Well if that's the case why not just use the `<audio>` tag? I don't think I understand how that is connected with the actual playing of the sound.

> store it in a separate singleton instance and just load the sounds as they're needed

let's ask ourselves what essentially we are trying to load or store of play or display. when do we need the audio data? when should we release it? what depends on it?

from what you have said I gathered that there are two primary stateful actions happening: managing the loading and releasing of audio data which will be played at some point; actually playing audio data in response to some action.

if an `<audio />` tag can be used to manage the loading of the audio data then that seems like a great plan because the browser handles everything the middleware would otherwise and it does it in native code. how then do we trigger the playing of that sound? how could we test those behaviors?

sorry I don't have answers prepared.

kriskarkoski commented:
It sounds like there may be some further changes from the above discussion, but as it is now it tested while for me in Chrome on Mac both with the chat open and collapsed behind the footer icon.

jordwest commented:
@dmsnell I've had a go at implementing it in middleware, and I'm convinced :) It's definitely nicer.

I didn't implement load/unload logic, since it's not super important that the sound plays at the precise moment it's needed. Instead the sound is just loaded at the time it's first needed.

I haven't had a chance to look at the issue of it not playing on IE yet, as the Windows VM seems to keep crashing my machine.

dmsnell commented:
@jordwest instead of commenting directly in the code I will just leave a more general comment here. I'm glad you are finding it cleaner. it's a good move. there's more to go over still though.

for starters, why are we storing audio DOM elements?

jordwest commented:
> for starters, why are we storing audio DOM elements?

The `new Audio()` JS API provides an simple API to play sounds, which is effectively the same as a `<audio>` tag from the browser's perspective. It does return an [`HTMLAudioElement`](https://developer.mozilla.org/en/docs/Web/API/HTMLAudioElement), but that object itself doesn't need to be appended to the DOM to be used, it's just an object with a `play()` property.

Adding an `<audio>` tag to the DOM is certainly another option, but I couldn't think of a clean way to do this because of the need to trigger a method on it. The imperative action of playing a sound is kind of in conflict with React's declarative nature. The ideas I've had thus far are:

 - Individual unique sound files could each have their own long-lived `<audio>` element appended to the DOM. Then we need to find a way to reach into that element and trigger its `play()` method each time the action fires. This could be done with refs, perhaps a component could watch the state, but that's getting a little complex.

 - Audio play actions can be recorded in the state tree, and React could render a corresponding `<audio>` element with an `autoplay` property for each of these. This would end up creating a new element each time the sound plays though, and we'd need to clean them up when they finish playing.

 - Other ideas?

dmsnell commented:
For starters, thanks for being so willing to explore the ideas!

One thought I had is that if we were to load an audio tag without playing it then the browser would automatically cache it. If we then used the Redux action to specify the URL for the sound then there might not be any real delay in completely recreating that object thus not needing to store it or purge it from memory

jordwest commented:
That's a nice idea, and in fact according to the docs we can immediately let the Audio element go out of scope:

> Note: An Audio element constructed with new Audio() won't be garbage collected so long as playback is in progress. It will continue playing and be audible until stop() is called or playback finishes.
> ([HTMLAudioElement Mozilla Docs](https://developer.mozilla.org/en/docs/Web/API/HTMLAudioElement))

That could mean we end up allocating/deallocating elements unnecessarily, but considering how infrequently the sounds will be played I don't think thats an issue.

jordwest commented:
I've simplified things in 4290e0d, removing the class that kept track of the Audio objects

omarjackman commented on client/state/audio/middleware.js:
```diff
@@ -0,0 +1,64 @@
+
+export const handlers = {
+	// Actions which trigger a sound
+	[ HAPPYCHAT_RECEIVE_EVENT ]: dispatchHappychatSound,
```
I wonder if this should be in it's own middleware. Would it be overkill to add another middleware specifically for happychat and include this trigger there? That way the `client/state/audio/middleware.js` middleware is only concerned with playing audio and doesn't care what audio it plays. Also I'm thinking in the future happychat might also want to do a desktop notification and that could also go in the same middleware.

dmsnell replied:
Good question @omarjackman! I could see it either way depending on how much logic is involved. In this case I'd probably lean on putting all of the audio-related actions into this middleware file since it will probably remain small and to-the-point.

@jordwest I may however recommend eliminating the separate function call and dispatch for HappyChat here. is there a need to do that? since the bell itself is a side-effecting response to another action I don't think we need to create yet another action.

```js
const playSound = soundName => () => {
	const url = get( soundNames, soundName );
	if ( ! url ) {
		return;
	}
	const soundClip = new Audio( url );
	soundClip.play();
}

export const handlers = {
	[ HAPPYCHAT_RECEIVE_EVENT ]: playSound( 'happychat-message-received' ),
}
```

I see now that I forgot a logic check…

```js
const playSound = soundName => …

const playHappyChatDing = ( { event } ) =>
	event.source !== 'customer' && playSound( 'happychat-message-received' );

export const handlers = {
	[ HAPPYCHAT ]: playHappyChatDing,
}
```

jordwest replied:
Both great points.

I agree that the `AUDIO_PLAY` action is a bit redundant here, so I've removed it in f9d76d7 which simplifies things further.

dmsnell commented on client/state/audio/actions.js:
```diff
@@ -0,0 +1,14 @@
+		type: AUDIO_PLAY,
+		reference,
+	};
+}
```
fun fact! we're allowed to simplify the syntax here since all we're doing is returning an object from the function

```js
export const audioPlay = reference => ( {
	type: AUDIO_PLAY,
	reference,
} );
```

dmsnell commented on client/state/audio/middleware.js:
```diff
@@ -0,0 +1,64 @@
+		return;
+	}
+
+	const audio = new Audio( REFERENCE_MAP[ reference ] );
```
let's be careful here about `undefined` property accesses. if the value doesn't exist in the map we shouldn't try to play it. `lodash.has()` and `lodash.get` can be good utilities to use. they default to only looking at enumerable properties (safer)

```js
const url = get( REFERENCE_MAP, reference );
if ( ! url ) {
	// error?
	return;
}

const soundClip = new Audio( url )
```

jordwest replied:
Calling `new Audio( undefined )` is valid and will just create an empty element, but you're right that we should be defensive here.

Fixed in 6ecbc05.

dmsnell commented on client/state/audio/middleware.js:
```diff
@@ -0,0 +1,64 @@
+	}
+
+	const audio = new Audio( REFERENCE_MAP[ reference ] );
+	audio.play();
```
a quick note about naming. `audio` seems fine in this small function, but it's not exactly descriptive. how about something a little more helpful like `soundBite` or `soundClip` or `dinger` haha

jordwest replied:
Renamed to `audioClip` in f9d76d7

dmsnell commented on client/state/audio/middleware.js:
```diff
@@ -0,0 +1,64 @@
+export default ( { dispatch, getState } ) => ( next ) => ( action ) => {
+	if ( handlers.hasOwnProperty( action.type ) ) {
+		handlers[ action.type ]( dispatch, action, getState );
+	}
```
One thing @aduth pointed out to me is the relative performance of objects with and without a prototype. It's faster here (since this runs on every dispatch) to start with a null prototype.

```js
const handlers = Object.create( null );
handlers[ HAPPYCHAT_… ] = …
```

further, and I'd appreciate @aduth's thoughts on this, the newer `Map()` datatype in JavaScript seems to have even faster read throughput than objects without prototypes.

```js
const handlers = new Map( [
	[ HAPPYCHAT_RECEIVE_EVENT, playSound( … ) ],
] );

if ( handlers.has( action.type ) ) {
	…
}
```

jordwest replied:
Are the performance benefits here worth the less readable code? I looked at `Map`, however browser support still seems a bit flaky.

Perhaps if perf is important here we could use `Immutable.Map` here instead which brings a consistent API and some performance benefits

dmsnell replied:
> benefits here worth the less readable code?

is there something that seems particularly hard to read here? I do recognize that it's a tradeoff but we must remember that this function gets called on every single dispatch throughout Calypso, so performance really is a big deal here.

`Immutable.Map` will be significantly more overhead than even an object lookup. Just creating a `null` object prototype than filling it out as a map isn't so different than what was here originally.

jordwest replied:
Replaced with `Object.create( null )` in 8a0032a.

It's still readable, but on first glance I'd be confused as to why it was initialized that way unless I knew it was a common performance hack. I've left a comment above explaining why it's done that way.

I'm curious where the theory comes from? At least in my [primitive benchmarks](https://gist.github.com/jordwest/8c884e57b2f056911ed68710d60e4b8a), they seem practically equivalent with only a difference varying on the order of hundreds of nanoseconds.

aduth replied:
>I'm curious where the theory comes from?

The optimization comes from the lookup operator. `hasOwnProperty` can be slow (I imagine `has` is even slower):

https://measurethat.net/Benchmarks/Show/1214/0/undefined-vs-typeof-vs-in-vs-hasownproperty

Specifically in the previous discussion Firefox was showing to be particularly slow on `hasOwnProperty`:

https://github.com/Automattic/wp-calypso/pull/8606#discussion_r92007182

The need for `Object.create` vs. `{}` is to avoid operators other than `hasOwnProperty` checking an object's prototype. `Object.create` creates an object without any prototype members (e.g. compare `'valueOf' in {}` vs. `'valueOf in Object.create( null )`).

This is both an edge case and a micro-optimization, but I'm particularly sensitive to middlewares being as fast as possible (because we dispatch so much). In fact, I'd much rather we have as _few_ middlewares as possible. And especially because this pattern (looking up handlers in an object) is becoming so common, I wonder if we should think about consolidating them.

jordwest replied:
Thanks for the info @aduth.

Is consolidating the middlewares something that should block this PR or can we leave it to a future PR?

jordwest commented:
Thanks for the reviews @omarjackman and @dmsnell!

I've made changes in response and left inline comments. I'd be grateful if you could take another look!

dmsnell commented on client/state/audio/middleware.js:
```diff
@@ -0,0 +1,62 @@
+		return;
+	}
+
+	const audioSource = getSourceForReference( reference );
```
Based on how this has changed with `get()` is it worth having a separate function for it?

jordwest replied:
Removed in 8a0032a

dmsnell commented on client/state/audio/middleware.js:
```diff
@@ -0,0 +1,62 @@
+};
+
+export const onHappyChatMessage = ( dispatch, { event } ) => {
+	event.source !== 'customer' && playSound( 'happychat-message-received' );
```
is having a map necessary since these references and files all appear to be contained in this single file? maybe with the added simplicity of having all the sounds tied to the actions in one place the mapping isn't as important, or at least, with only a single sound it isn't import yet?

jordwest replied:
Good point, I figured this would be extensible if in future we need to change the reference, but it's probably premature. Removed the references and map in 8a0032a.

dmsnell commented on client/state/audio/middleware.js:
```diff
@@ -0,0 +1,45 @@
+	if ( typeof Audio !== 'function' ) {
+		// No Audio support in this browser
+		return;
+	}
```
I think I lost a comment here so I'm going to repeat it and hope I'm not duplicating/nagging you.

what is actually going on here if this condition evaluates false?

we're not actually wanting to do anything and it's not like we boot up the app _with_ audio support then at some point in the runtime lose it.

while this is a great check _to keep in here_ I think we could benefit by putting two additional checks purely for safety. one at the point where we actually inject the middleware (no need to inject it if no `Audio` is available) and the other is in the default export below. if for some reason we accidentally inject it we can skip initializing the middleware completely.

@blowery has redone the middleware chain in #10897 and this should probably wait to rebase off of those changes

jordwest replied:
Cool, thanks for the heads up. I'll add the conditional injection once #10897 is merged.

jordwest commented:
@dmsnell I've added the conditional middleware injection in the rebase, and added the additional check in the middleware function as suggested in 1c551dd.

Could you take another look? Thanks again for your ongoing review :)

dmsnell commented:
@jordwest for me there remains two things I would like to see before this merges:

 - replace the `has()` with something faster (see my comment)
 - README? since this introduces an entirely new middleware we should be able to tell people _why_ it exists and _how_ to use it. please include some examples and rationale for how new additions should be made - high-level stuff vs. a straight listing of functions and parameters (that's what JSDoc is for)

jordwest commented:
@dmsnell Good suggestions!

 - Replaced `has` in 308fbd9
 - Added README in 8b18b3f
 - I've also added a few unit tests (which caught a failure to call `next`) in c126f21

dmsnell commented on client/state/audio/middleware.js:
```diff
@@ -0,0 +1,46 @@
+export default ( { dispatch, getState } ) => ( next ) => ( action ) => {
+	if ( ! isAudioSupported() ) {
+		return next( action );
+	}
```
by putting this check here it will run on every action. however, since this is a curried function we can actually move it up a few levels and then it will only run once (in my example code I'm going to make it a named export so that we can expose it explicitly to tests)

```js
export const middleware = ( { dispatch } ) => next => {
	if ( ! isAudioSupported() ) {
		return next;
	}

	return next => {
		const handler = handlers[ action.type ];
		if ( handler ) {
			handler( dispatch, action );
		}

		return next( action );
	};
}
```

dmsnell commented on client/state/audio/middleware.js:
```diff
@@ -0,0 +1,46 @@
+
+	const handler = handlers[ action.type ];
+	if ( 'function' === typeof handler ) {
+		handler( dispatch, action, getState );
```
do we really need to pass in `getState`?

I'm okay with it, but it doesn't look like we use it, and in the audio middleware so far it isn't being used, and I'm not sure I see an obvious case where it would be necessary

dmsnell commented on client/state/audio/middleware.js:
```diff
@@ -0,0 +1,46 @@
+
+// Initialized this way for performance reasons
+export const handlers = Object.create( null );
+handlers[ HAPPYCHAT_RECEIVE_EVENT ] = onHappyChatMessage;
```
non-blocking comment on naming here: does `onHappyChatMessage` tell us much? we generally want to avoid `onXYZ` names anyway. but what is the purpose of this function?

```js
handlers[ HAPPYCHAT_RECEIVE_EVENT ] = playDingOnCustomerMessages;
```

??

dmsnell commented on client/state/audio/middleware.js:
```diff
@@ -0,0 +1,46 @@
+	HAPPYCHAT_RECEIVE_EVENT,
+} from 'state/action-types';
+
+const isAudioSupported = () => typeof window === 'object' && typeof window.Audio === 'function';
```
😻

jordwest replied:
It was that way before, but it makes it harder to test

gwwar replied:
Would this ever change in a single page view? If you need to stub it in tests you can do something like the following or stub the window.Audio property.

```js
export const isAudioSupported  = typeof window === 'object' && typeof window.Audio === 'function';
```

gwwar replied:
This is a bit messy to stub.  Another alternative is to export an object. You can then update the reference there without checking these props or manipulating the window object.

```jsx
const export isSupported = { audio: typeof window === 'object' && typeof window.Audio === 'function' };
```

Then in the test you can do:
```
import middleware, { isSupported } from '../middleware';
// In your test setup
isSupported.audio = false;
```

dmsnell replied:
of note, we've had plenty of discussion on creating a "browser object model" in Calypso which we could use for all browser-related functionality and still provide for easier testing and mocking. just a thought.

dmsnell commented on client/state/audio/test/middleware.js:
```diff
@@ -0,0 +1,84 @@
+			dispatch: spy(),
+			getState: stub(),
+			replaceReducers: spy(),
+			subscribe: spy(),
```
I won't put my foot down against this but do we really need to mock the full store since these functions don't use anything other than `dispatch`?

we can test to make sure that nobody introduces code changes which do this though I would hope that our review process will catch any of those things. they would be odd anti-patterns anyway if someone used the audio middleware to replace the reducers though maybe not technical bugs

dmsnell commented on client/state/audio/test/middleware.js:
```diff
@@ -0,0 +1,84 @@
+			subscribe: spy(),
+		};
+
+		store.getState.returns( Object.create( null ) );
```
can probably just skip all these. the tests don't even concern themselves with them.

dmsnell commented on client/state/audio/test/middleware.js:
```diff
@@ -0,0 +1,84 @@
+			Audio: spy( function() {
+				return { play };
+			} ),
+		};
```
cool

dmsnell commented on client/state/audio/test/middleware.js:
```diff
@@ -0,0 +1,84 @@
+	} );
+
+	afterEach( () => {
+		global.window = undefined;
```
I trust you if you say yes, but are we sure this is the right way to reset `global` after the tests?

dmsnell commented on client/state/audio/test/middleware.js:
```diff
@@ -0,0 +1,84 @@
+		expect( next ).to.have.been.calledWith( action );
+	} );
+
+	it( 'should fail silently when no audio support', () => {
```
"fail"? how about "should not attempt to play sound when no audio support in browser"

dmsnell commented on client/state/audio/middleware.js:
```diff
@@ -0,0 +1,46 @@
+};
+
+export const onHappyChatMessage = ( dispatch, { event } ) => {
+	event && event.source !== 'customer' && playSound( '/calypso/audio/chat-pling.wav' );
```
so I was confused on which messages should trigger a ding. I think naming could help out a bit here. something like `playNewOperatorMessageDing` (probably need a better name than that) maybe

dmsnell commented on client/state/audio/test/middleware.js:
```diff
@@ -0,0 +1,84 @@
+		expect( play ).to.not.have.beenCalled;
+	} );
+
+	it( 'should play sound on HAPPYCHAT_RECEIVE_EVENT action', () => {
```
how about `should play sound for customer when receiving a new message from the operator` or some message to convey this?

dmsnell commented on client/state/index.js:
```diff
@@ -119,13 +119,15 @@ const atEnabled = isDevOrWPCalypso && config.isEnabled( 'automated-transfer' );
 
 export function createReduxStore( initialState = {} ) {
 	const isBrowser = typeof window === 'object';
+	const isAudioSupported = typeof Audio === 'function';
```
just a note that this check is different than the one inside the middleware where we test that we're in a window _and_ that `Audio` is a function

dmsnell commented on client/state/audio/README.md:
````diff
@@ -0,0 +1,36 @@
+
+handlers[ SOME_ACTION_TYPE ] = onSomeEvent;
+
+```
````
this code snippet is a bit confusing to me. do you think we could get a little more context here? maybe add two example events, one of which performs some logical test, one that doesn't? I think adding the HappyChat function could be a good example.

how about also putting the middleware filename as a comment at the top of the example block? help us orient ourselves in the middleware

jordwest commented:
Thanks for the detailed review as always @dmsnell. I've taken on board most of your suggestions in 8b90489, with the exception of adding an extra example in the README (I don't think it's worth copying the Happychat example into the README since it's already in the code)

dmsnell commented:
I haven't tested yet but all looks good to me otherwise :+1:

jordwest commented:
Rebased to fix conflicts with master, retested.

@dmsnell are you happy for this to go in as is?

dmsnell commented:
:shipit:

gwwar commented:
No chat operators were on, but I smoke tested this a bit in a local Docker 🐋 env. Things look good. Was able to post make a new site, etc.

jordwest commented:
Thanks @dmsnell and @gwwar! 

I've given operators a heads up (p7DVsv-1OL-p2) and will merge this on Monday
